### DISCUSSION TITLE
Creating QCodes Driver with existing Python Interface

### DISCUSSION THREAD

**Discussion by Diddleydope** [microsoft/Qcodes#5765] [Q&A]
Hello everyone!

I just finished creating a Python TCP interface for our instruments, which connects directly via TCP and allows users to interact with our instruments trough Python. There is significant demant for a QCodes Driver for our devices, however we do not support VISA, so I thought to use the IPInstrument class instead. Here is where I've run into my first problems. I want to integrate the Python interface into the package, which already has type checking, parsing and everything implemented for all the device's functions. I am unsure of two things: Firstly, should I use Parameters or normal functions to define all of the functions from the Python interface in the QCodeS driver? (see picture, nanonis_specs_specs is my python interface)
<img width="673" alt="Screenshot 2024-02-23 at 10 55 36" src="https://github.com/microsoft/Qcodes/assets/72755021/dcc01a65-a81e-49fb-8813-e02a73983ad0">

Secondly, the way I have the Parameter right now, it's not working. It is unclear to me how i tell the parameter how many parameters my get or set cmd has. I find the documentation to be a bit unclear. The issue I have with just defining a bunch of functions, is that I have to give the connection to the python interface as a parameter (see picture). I would appreciate any help on the matter.

Edit: 

For clarification, this is the error I get when trying to call the parameter:

"TimeoutError: ('timed out', "writing '<nanonis_spm_specs.NanonisClass.Nanonis object at 0x120032600>.returnDebugInfo' to <NanonisSPM: NanonisSPM>", 'setting NanonisSPM_returnDebugInfo to 1')"

<img width="825" alt="Screenshot 2024-02-23 at 12 27 01" src="https://github.com/microsoft/Qcodes/assets/72755021/8e96e65b-36ca-4516-b125-5307e274da44">

Kind regards,

Samuel O'Neill, SPECS Zurich

**Comment by jenshnielsen** [comment]
@Diddleydope Did you get your issue resolved (I can see that you marked this as such)

**Comment by Diddleydope** [comment]
Dear Mr Nielsen,

The issue is not resolved per se, but we've found the root of the problem.
Our python interface formats the data sent over TCP fundamentally
differently than the IPInstrument class does. Our data is sent and parsed
as a bytearray with a set size header, which defines the size of the body,
and the data is parsed accordingly in the package. So the problem is, that
the parameter in QCodeS is executing our function that I gave it as a
set_cmd, but is executing it through the TCP connection in the python
interface, which returns a bytearray and not a string, which is what the
IPInstrument class needs/expects. Hence the function WAS being executed,
but I was getting an Error from the recv function in IPInstrument, because
it was getting a bytearray. And since we can't modify the way the
IPInstrument class sends/receives data, we've run into a dead end. Perhaps
a possibility would be to inherit from Instrument and make a
SPECSInstrument Class, where we implement our own instrument specific
send/receive methods to properly parse our data or just call our own python
interface.

**Comment by jenshnielsen** [comment]
Ok thanks for the update. You may want to simply not use the IP Instrument class but directly subclass Instrument to implement the logic that you need. The main things that you need are to implement `get_raw` and `set_raw` in a way that matches how you would like to communicate with your instrument.
